Menu styling rework

brettsmason opened the pull request:
This is an attempt at starting a rework for the current menu system. See #9421.

## Aim of this PR

- To reduce the specificity of the current `.menu` class and make it more flexible
- Remove `table-cell` from `.menu` (as its a nightmare to work with!)
- Add more styling flexibility to each of the menu plugins (dropdown/drilldown/accordion) by adding new variables to allow for more control

## Current Problem

`.menu` includes its own padding values, along with heavy use of `>` which makes it very difficult to override with your own styles. I suggest making `.menu` almost like a base reset for menus. Then other classes like `.simple` or `dropdown` add their own margin/padding and there's no issue with overriding the inherited `.menu` styles.

This PR is still very early on and am experimenting with it a lot. There is a visual test under menu that will show you the various classes and how they work with each other.

brettsmason committed:
Start the rework of the menu system. Added a visual test to help work on these styles.

brettsmason committed:
Update and organise visual test

brettsmason committed:
First pass at making the menu plugin styles more flexible.

brettsmason committed:
Correct missing & on alignment classes.

brettsmason committed:
Set correct dropdown menu background colour

brettsmason committed:
Add menu icon styles - still needs refining. Added some examples to the visual test.

brettsmason committed:
Add option to allow simple vertical menus. Needs revisiting to be less specific.

brettsmason committed:
Add back menu-text, though needs reworking a bit. Added responsive expanded classes, so you can have this behaviour on large and up for example.

brettsmason committed:
Add a `$global-menu-padding` variable. This sets the padding for all menu components, however it can be changed for each component by altering their own padding values eg `$accordionmenu-padding`. This should allow greater flexibility.

brettsmason committed:
Change the order of `.expanded` to allow it to work correctly with orientation classes. Only target top level <li> elements so it works correctly with dropdown menu.

brettsmason committed:
Add test to show that inputs, select and buttons work correctly within a menu.

brettsmason committed:
Add `display: block` on <a> elements by default, and remove from direction mixin.

IamManchanda commented:
awesome @brettsmason !!!

brettsmason committed:
Add flexbox version of the test to allow for testing on flexbox specific styling.

brettsmason committed:
Fix missing display: block for flex items.

brettsmason committed:
Remove some flexbox styles that were overly opinionated.

brettsmason commented:
@IamManchanda Thanks! Not to worried about conflicts at the moment as its very much work in process and an exploration of what we can do to improve the menu system.

@kball @rafibomb I think this should be ready for me to discuss with you guys on Thursday assuming that's still happening. I have a few more ideas, but probably best to check with you first.

brettsmason committed:
Fix for flexbox version of accordion submenu not being full width.

brettsmason committed:
Reorganise output classes.

kball commented:
@brettsmason checking in real quick - what is the state of this?

brettsmason commented:
@kball I think from memory it was in a state were it could now be tested to see how it works. I think from a previous video chat you guys were going to test it in a project to see if it worked OK. 

I'll double check though as it's been a while.

IamManchanda commented:
Menu with input, select and buttons is having visual issues - More with flex version
Rest looks good visually!

kball commented:
@brettsmason or @IamManchanda can either of you look at fixing the issues highlighted above before the first 6.4 RC (End of May)?

IamManchanda committed:
Fix conflicting padding between native button and anchor button!

IamManchanda committed:
Add minimal margin-left and right for flexbox code

IamManchanda commented:
Kindly check now @kball .... i have updated 
and now the output....

![menu-flex-styling](https://cloud.githubusercontent.com/assets/4970624/26258204/42440a86-3ce1-11e7-92c9-afb79f69cbac.png)

brettsmason commented:
I think this could still do with a full review before it gets merged, as far as I know it's only been quickly looked at. I'll take another look to make sure I finished everything - lost motivation on it a bit as nobody looked at it! :)

IamManchanda commented:
> lost motivation on it a bit as nobody looked at it

Dang .... dont loose motivation as harry is on it!

IamManchanda commented:
@brettsmason @kball how about creating responsive breakpoints for accordion's nested class 
https://github.com/zurb/foundation-sites/issues/10002 https://codepen.io/wecodeart/pen/xddBZz

brettsmason commented:
@IamManchanda I'm not sure I see the need for that? If you look at the new code there are now variables and styling to control individual navigation style components (eg `.accordion-menu`). I personally think thats enough control, as I was trying to cut back on the bloat 😄

IamManchanda commented:
@brettsmason But we have that one way to minimize that bloat!

```scss
/// Responsive breakpoints for nested menu
/// @type Boolean
$responsive-menu-nested: false !default;
```

```scss
.menu {
  // Nesting
  &.nested {
    @include menu-nested;

    @if ($responsive-menu-nested) {
      // Loop through Responsive Breakpoints
      @each $size in $breakpoint-classes {
        @if $size != $-zf-zero-breakpoint {
          @include breakpoint($size) {
            &-for-#{$size} {
              @include menu-nested;
            }
          }
        }
        @include breakpoint($size only) {
          &-for-#{$size}-only {
            @include menu-nested;
          }
        }
      }
    }
  }
}
```

IamManchanda commented:
@BicanMarianValeriu being the author of that nested thread ..... What do you think ??

BicanMarianValeriu commented:
@IamManchanda 
You know what I think :)). It looks really ugly by default if nested class is used (mainly for vertical menus) and the margin will also remain for dropdowns. Another way of doing this is to remove that left margin of the nested class for horizontal dropdowns.

We can do that with: 
```
.nested.is-dropdown-submenu {
// When menu turns to accordion this class is transformed to is-accordion-submenu
// and we will still have that margin
margin-left:0;
}
```

I think this is better, not that much bloat but test it, you know better.

brettsmason commented:
@IamManchanda Looking at this properly now, I think we should add a new variable to each 'mobile' menu component (drilldown, accordion) eg: `$accordionmenu-nested-margin: null !default;` (above here: https://github.com/brettsmason/foundation-sites/blob/683b1c6639a00008373c9af77711b87bf8b9aabe/scss/components/_accordion-menu.scss#L11)

This would allow control over if the menus should have margin on them or not (null by default so nothing is output). Then a style needs adding for `.submenu` with the value of this variable.

Would that solve this do you think? I think its going to be the cleanest way.

brettsmason reviewed (commented):
Left a quick bit of feedback (added this by mistake 😆 )

brettsmason commented on scss/components/_menu.scss:
```diff
@@ -52,6 +52,11 @@ $menu-item-background-hover: $light-gray !default;
   li {
     @include disable-mouse-outline;
+
+    @if $global-flexbox {
```
Not sure I understand the point of this, could you explain? If we are adding spacing between `li` items, it should be at least controllable by the user. Why not have them sitting flush together though?

IamManchanda replied:
Yeah sitting flush together also would be good... I was wanting it to similar to float!

brettsmason replied:
Remember flexbox is going to be primary from 6.4, and the space between in is more a bug with `inline-block`. I recommend we remove the margin and sit flush personally, as it would generally be better (and in 7 there would be no float anyway I expect!).

IamManchanda replied:
> and in 7 there would be no float anyway I expect

Then how would you get your goverment clients :sweat_smile: who uses IE all the time :wink: 
Ohk removing it!

brettsmason replied:
I'll have to stick to V6 for them 😢

IamManchanda replied:
:smile: :smile_cat: :smiling_imp: :joy: :stuck_out_tongue_closed_eyes:

IamManchanda commented:
@brettsmason :+1: 

But man looking at @BicanMarianValeriu ,
I was looking at drilldown accordion combo

```scss
.nested.is-dropdown-submenu,
.nested.is-drilldown-submenu{
   margin-left:0;
}
```

and man i found another bug :sweat_smile: 
=> https://codepen.io/IamManchanda/pen/BRGJGL
Similar to #10031... Gonna report it :smile:

IamManchanda committed:
Added a test case - Vertical Menu Alignment

brettsmason commented:
>  I was advancing and looking at drilldown accordion combo

That's why I said we would have the same in the drilldown partial 😄 It would be cleaner, and the value would be totally configurable for Sass users. And for CSS users, its not hard to write the one line needed to add margin only on accordion for example (in my opinion) 👍

IamManchanda commented:
@brettsmason Related to our discussion in #9985 .... i added a test case 
(which is currently not working ... as we can see) .... How can we best solve this??
(Without conflicting the specificity part ofcourse)

brettsmason commented:
@IamManchanda Thanks for putting the test case together. I've been wrapped up in the new grid, but this evening I'll pull down this menu rework and familiarise myself with it again so I understand it properly again 😄 Then I will be able to help more 👍

IamManchanda committed:
Revert: Let li sit flush together with each other in flexbox!

IamManchanda commented:
@brettsmason ..... As discussed taking this up .... best of luck for the grid :wink:

IamManchanda committed:
Solve #9985 as per specificity!

IamManchanda committed:
Left align the submenu of the dropdown - `flex-start`

IamManchanda committed:
Test case for vertical dropdown menu!

IamManchanda committed:
Closes #10002 - Fixes responsive menu nested bug

Changes:
1. Nested margin
a. Global Variable
- Adds a Global Variable `global-menu-nested-margin`, 1 rem by default
b. Simple Menu
- `menu-nested-margin` defaults to `global-menu-nested-margin`
- `.nested` gets a mixin `menu-nested` with default arguments
c. Accordion Menu
- `accordionmenu-nested-margin` defaults to `global-menu-nested-margin`
- `.nested` gets a mixin `menu-nested` with `accordionmenu-nested-margin` as an argument
d. Dropdown Menu
- `dropdownmenu-nested-margin` defaults to `global-menu-nested-margin`
- `.nested` gets a mixin `menu-nested` with `dropdownmenu-nested-margin` as an argument
e. Drilldown Menu
- `drilldown-nested-margin` defaults to `global-menu-nested-margin`
- `.nested` gets a mixin `menu-nested` with `drilldown-nested-margin` as an argument

2. Drilldown Nesting Fixes
3. Added visual test case for nesting!

IamManchanda committed:
Changed 0px to 0

IamManchanda commented:
Hey @brettsmason, me and @kball had discussion about #10002
Although we both liked your [idea](https://github.com/zurb/foundation-sites/pull/9670#issuecomment-302920237)

But the problem was Nested class is good for simple menu and accordion menu 
but ugly for dropdown and drilldown. 
And thus was bad responsive menu!

So here's what we got up to 

1.  Added a global variable 

```scss
$global-menu-nested-margin: 1rem !default;
```
2. Define simple menu and accordion menu the global variable 

```scss
$menu-nested-margin: $global-menu-nested-margin !default;
$accordionmenu-nested-margin: $global-menu-nested-margin !default; 
```

3. Define dropdown menu and drilldown menu to `0` .... so no nesting gets here!

```scss
$dropdownmenu-nested-margin: 0 !default; 
$drilldown-nested-margin: 0 !default; 
```

Now i added the classes like these 

```scss
.accordion-menu {
  /* my other code */
  .nested.is-accordion-submenu {
     @include menu-nested($accordionmenu-nested-margin);
   }
}
```
and same for other two! Sorry if you dont concur with this. 
Let me know your thoughts!

brettsmason commented:
@IamManchanda That sounds a really good plan 👍 

Just a few comments:
The `$global-menu-*` variables - should this be `$menu-global-*` instead? Would sit better under the list of variables, and other variables are all prefixed with their component name.

> .nested.is-accordion-submenu {

Cant this just be `.nested`? No need for the extra specificity unless I'm missing something.

IamManchanda commented:
@brettsmason 
> That sounds a really good plan 👍 

Not a plan now as implemented :smile: 

> The `$global-menu-*` variables - should this be `$menu-global-*` instead? Would sit better under the list of variables, and other variables are all prefixed with their component name.

All the variables in the global file goes with the naming convention like this.
`$global-radius` , `$global-menu-padding`, `$global-flexbox`
So i have no doubts in my mind that it should be this covention 

> > .nested.is-accordion-submenu {
>
> Cant this just be `.nested`? No need for the extra specificity unless I'm missing something.

@kball what do you think on this .... on just being `.nested` ??

IamManchanda committed:
Tweaks to vertical menu alignment (float)

IamManchanda committed:
Test case for vertical menu alignment (float)

IamManchanda committed:
Add a non working buggy test case!

kball commented:
Do we ever have non-accordion menus nested within accordion menus? If so, I think we need the extra specificity.

IamManchanda committed:
Add test case for advanced nested menu

IamManchanda commented:
@kball We actually have it ..... yes menu nested inside a menu .... just added a test case for the same above :arrow_up:  

And this extra specificity only helping it unless it will break! 

![drilldown-accordion-nested-menu](https://cloud.githubusercontent.com/assets/4970624/26421026/210a3dd6-40e2-11e7-9151-9a426cfbe5db.png)

brettsmason commented:
@IamManchanda So I've been trying to catch up a bit on what's been going on with menu.
Just for the record I still disagree with the `.nested` stuff 😄 I don't understand why you would need to use `.nested` in the situations I've seen mentioned. Is it this that you need help with? Cant quite work it out...

IamManchanda commented:
There will be disagreement.... 
All I can say is you need `.nested` for simple and accordion menu 
But with drop-down and drill down, you don't need that styling and it looks ugly

IamManchanda committed:
Updated test for right alignment dropdown menu!

Also Shelved vertical central menu as unuseful

IamManchanda commented:
@brettsmason As discussed, i have updated test case for right aligned vertical dropdown menu. Please look into it now!

For central menu, As discussed .... as unuseful and was just a test that didn't work .... i have shelved it.
**Marking it as wont fix**!

IamManchanda committed:
Added test case for buggy right aligned vertical accordion menu!

IamManchanda committed:
Add test case for buggy drilldown vertical aligned right menu!

IamManchanda commented:
@brettsmason @abdullahsalem @kball I am able to reproduce the same right aligned vertical issues in accordion and drilldown too . Please help i added test case too 

Here are buggy test path
/menu/vertical-dropdownmenu-alignment.html
/menu/vertical-dropdownmenu-alignment-flex.html
/menu/vertical-drilldownmenu-alignment.html
/menu/vertical-drilldownmenu-alignment-flex.html
/menu/vertical-accordionmenu-alignment.html 
/menu/vertical-accordionmenu-alignment-flex.html

PS: Sorry i know i made lot of test cases :p .... but will push it to `all-menus.html` once we solve this.

kball committed:
Fix dropdown menu arrow alignment

kball committed:
Fix drilldown menu arrow alignment

kball committed:
Fix accordion menu arrows

kball committed:
Merge branch 'develop' into better-menu-rework

kball commented:
I updated this PR to fix all of the arrows for right aligned case to at least pick the correct "side", and it now works nicely for the float-based menus, but I think there's a fundamental problem with the current approach for flex menus, at least for the non-simple menus (and possibly throughout), that manifests for vertical menus that have right alignment (actually global right, but I'm going to use "right" for simplicity here)

I have an idea of how to fix, but not sure it's the right thing, so want to get opinions...

The issue is that for vertical left-align menus, `a` tags are `display: block` inside a `display: block` `li` parent, which results in them being full-width, which 1. makes the whole row clickable, and 2. makes the arrow position correctly.

For vertical right-align menus, currently in flex mode we're making the `li` parent a flex parent, and using flex alignment to align the contents to the right. The problem with this is it makes the `a` tag compress to the size of the text, meaning the click target is only the text itself, and the arrow (which is attached to the `a`) positions oddly.

I think the solution is to not make the `li` a flex parent at all, and simply use `text-align` to position the text. However, I don't know the original reasoning for making the `li` a flex parent, so want to lay this out and see what @brettsmason and @IamManchanda think

andycochran commented:
@kball: I've suggested before that maybe arrows should just be inline-block and immediately adjacent to the text. They don't need to be aligned to the the global right of the menu item. It's actually bad UX if they are aligned to the right of menu items that are expanded—'cause the arrow is closer to the follow button's text than it is to its own. Lemme see if I can track down that old comment… 

andycochran commented:
The arrow issue is pointed out here: https://github.com/zurb/foundation-sites/issues/9421#issuecomment-271967786 
and here: https://github.com/zurb/foundation-sites/issues/7095

kball commented:
@andycochran interesting... that makes complete sense to me for horizontal menus, but for vertical it feels like we should expand to fill the width of the menu

IamManchanda committed:
More Fix for float(only) - vertical align right menus

IamManchanda committed:
Correct test case!

IamManchanda commented:
@kball .... I added it here https://github.com/zurb/foundation-sites/pull/9670/commits/1a5a6f2f63fe3e2cb0464a41860cd60dfd2354ab
I used to solve it to create vertical menu actually for flexbox.
If i remove flex from there ..... vertical `align-right, align-center` menu fails in flexbox actually (See `/menu/vertical-menu-alignment-flex.html` for testing )
Maybe we can do it in `.submenu li`?? 
But i am not sure

IamManchanda commented:
@brettsmason BTW ..... are we dropping `menu-centered` ??

brettsmason commented:
@IamManchanda If `.align-center' does what we need, then no I dont think its needed :thumbsup:

IamManchanda commented:
@brettsmason Same i feel .... but i am amazed we didn't add this in migration/depreciation ??
https://github.com/zurb/foundation-sites/wiki/Project-Roadmap#64-tentative-date-may

kball commented:
@IamManchanda we've started following a diff migration tracking approach (each PR at the end around merging if it requires a migration adds 'Migration Needed' tag, with comment about how the migration works... that way at the end we can pull them all up and note them)

With regards to the align-center, align-right... can we simply have a vertical menu with those classes apply text-align?

IamManchanda committed:
Fix Align Right Vertical Menu - Flexbox!

IamManchanda commented:
Done @kball ..... Your suggestion .... yup works now!

IamManchanda committed:
Correct sass docs!

IamManchanda committed:
Fix vertically aligned right dropdown menu

IamManchanda committed:
Add nested class condition for right alignment!

IamManchanda committed:
Update docs on `menu-centered` => `align-center`

IamManchanda committed:
Update vertically align right menu docs!

kball commented on docs/pages/menu.md:
```diff
@@ -64,27 +64,21 @@ By default, each item in the menu aligns to the left. They can also be aligned t
 <br>
 
-To align items in the middle, add a wrapping element with the class `.menu-centered`.
+To align items in the middle, use `.align-center`
```
Add it to what? The `.menu`?

IamManchanda replied:
Done!

kball commented on docs/pages/menu.md:
````diff
@@ -64,27 +64,21 @@ By default, each item in the menu aligns to the left. They can also be aligned t
 </div>
 
 ```html_example
-<div class="menu-centered">
````
What will happen for folks migrating from previous versions who have "menu-centered". Will it work? Or not do anything?

kball commented on docs/pages/menu.md:
```diff
@@ -142,6 +136,39 @@ Add the `.vertical` class to a Menu to switch its orientation.
+<div class="warning callout">
+  <p>
+    The above vertically left &amp; right aligned menu are supported in all types of menu's. <br>
+    But as we don't think `align-center` is useful for dropdown, accordion or a drilldown menu, we have just kept it on basic menu only.
```
I think stating what we think is useful is not necessary... lets just state that `align-center` just works in basic menus.

IamManchanda replied:
Done!

kball commented on scss/components/_menu.scss:
```diff
@@ -34,40 +38,29 @@ $menu-icon-spacing: 0.25rem !default;
-    display: block;
-    padding: $menu-item-padding;
+  a,
+  .button {
```
All links and buttons inside a menu are this way? That feels a little heavy-handed... though I suppose this is low enough specificity we can override if we need to... @brettsmason what do you think?

brettsmason replied:
I think this is carried over fro the old menu system. You are right though, it maybe a bit heavy handed. Having said that, I get why its there - a mini reset. I think `text-decoration` could come out, as thats definately not needed (it should respect whatever settings you have). We could try removing it and seeing how this effects an example including buttons, but I think its needed as we dont use a class on our `a` links within menus to style them (and we needed to keep backwards compatibility).

kball commented on scss/components/_menu.scss:
```diff
@@ -218,7 +273,7 @@ $menu-border: $light-gray !default;
   }
   @else if $position == top {
-    > li > a {
+    li a {
```
Why do we drop the `>` specifications? This feels likely to apply too broadly now...

brettsmason replied:
Generally the aim of the whole PR was to cut down on specificity (`.menu > li > a` was a nightmare to change).
I think the way we are generating the classes, so the parent for that will be `.menu.icon-right` would make this specific enough, bit I'm open to suggestions or correction 😄

kball commented on scss/components/_menu.scss:
```diff
@@ -231,11 +286,35 @@ $menu-border: $light-gray !default;
+    }
+  }
+  @else if $position == bottom {
+    li a {
```
Same question as above on not havint `>`

kball reviewed (commented):
This is looking like a good step forward and I think we're getting close... some questions added, and I'd like to get opinions from @andycochran, @ncoden, and @brettsmason on this as well if possible.

IamManchanda commented:
@kball With the current code `.menu-centered` will not work!

IamManchanda commented:
@brettsmason Can you revert on removing `>` ?

IamManchanda committed:
Update docs as per contributors review!

IamManchanda committed:
Remove centered dropdown menu test case as we only support basic menu

IamManchanda committed:
Add test case on nested forms within menu!

kball committed:
Merge branch 'develop' into brettsmason-menu-rework

kball committed:
Update tests for forms inside menus

kball committed:
Update dropdown active states to override menus by default

kball commented:
OK, I'd love to redo the JS side of the menus as well, but we're out of time for something that complex.  Will have to tackle that for 6.5.

I think this is good to go! Going to merge.

With regards to migration, the one item I'm seeing right now is the old `.menu-center` approach has been removed... is there anything else?

kball commented:
Do we want to include for backwards compatibility a `.menu-centered > .menu` style? Just ran into a visual test broken b/c it hadn't been updated here, but we might want to create this option and deprecate for 1 release
